src/Storages/NATS/NATSProducer.cpp

```diff
@@ -15,32 +15,60 @@ static const auto MAX_BUFFERED = 131072;
 
 namespace ErrorCodes
 {
-    extern const int CANNOT_CONNECT_NATS;
     extern const int LOGICAL_ERROR;
 }
 
-NATSProducer::NATSProducer(
-    const NATSConfiguration & configuration_,
-    const String & subject_,
-    std::atomic<bool> & shutdown_called_,
-    LoggerPtr log_)
+NATSProducer::NATSProducer(const NATSConfiguration & configuration_, BackgroundSchedulePool & broker_schedule_pool_, const String & subject_, LoggerPtr log_)
     : AsynchronousMessageProducer(log_)
-    , connection(configuration_, log_)
+    , configuration(configuration_)
+    , event_handler(log)
     , subject(subject_)
-    , shutdown_called(shutdown_called_)
     , payloads(BATCH)
 {
+    /// One looping task for all consumers as they share the same connection == the same handler == the same event loop
+    looping_task = broker_schedule_pool_.createTask("NATSProducerLoopingTask", [this] { event_handler.runLoop(); });
```

Comment:
#### Yes, that's right and that's the main problem.

Libuv is not a thread-safe library, so the nats library registers a thread in which ```uv_run``` will be called, in order to call libuv methods directly or via an asynchronous callback.

<details>
  <summary>natsLibuv_SetThreadLocalLoop documentation</summary>

```C++
/** \brief Register the event loop with the thread running `uv_run()`.
 *
 * Since `libuv` is not thread-safe, the adapter needs to know in which
 * thread `uv_run()` will run for the given `loop`. It allows the adapter
 * to schedule events so that they are executed in the event loop thread.
 *
 * @param loop an event loop.
 */
void natsLibuv_SetThreadLocalLoop(uv_loop_t *loop)
```
</details>

The library also requires that the connection to the server occurs from the events loop thread. Therefore, when we are connecting, the event loop is not processed.
<details>
  <summary>natsLibuv_Attach sources</summary>

```C++
natsStatus natsLibuv_Attach(void **userData, void *loop, natsConnection *nc, natsSock socket)
{
    uv_loop_t       *uvLoop = static_cast<uv_loop_t*>(loop);
    bool            sched   = false;
    NATSLibuvEvents *nle    = static_cast<NATSLibuvEvents*>(*userData);
    natsStatus      s       = NATS_OK;

    sched = ((uv_key_get(&uvLoopThreadKey) != loop) ? true : false);

    // This is the first attach (when reconnecting, nle will be non-nullptrptr).
    if (nle == nullptr)
    {
        // This has to run from the event loop!
        if (sched)
            return NATS_ILLEGAL_STATE;

        ....
        /// uv_async_init - non-thread-safe function
        if ((s == NATS_OK) && (uv_async_init(uvLoop, nle->scheduler, uvAsyncCb) != 0))
        .....
    }

    .....
}
```
</details>

In the previous implementation, function ```natsLibuv_SetThreadLocalLoop``` was called from diffrent thread, even from those where ```uv_run``` was not called. For example in threads where we reconnect to server or a random thread from the background pool when we switched to it after rescheduling

#### I agree that this is wrong from a performance point of view, but perhaps this is the only correct way to use the nats library.

I tried to make one common event loop for consumers and producers, but I ran into the problem that the application terminates in the libuv library in the uv__epoll_ctl_flush function if the event loop freezes for a while, and we get ```cqe->res == -EBADF```
<details>
  <summary>uv__epoll_ctl_flush sources</summary>

```C++
static void uv__epoll_ctl_flush(int epollfd,
                                struct uv__iou* ctl,
                                struct epoll_event (*events)[256]) {
  struct epoll_event oldevents[256];
  struct uv__io_uring_cqe* cqe;
  uint32_t oldslot;
  uint32_t slot;
  uint32_t n;
  int fd;
  int op;
  int rc;

  STATIC_ASSERT(sizeof(oldevents) == sizeof(*events));
  assert(ctl->ringfd != -1);
  assert(*ctl->sqhead != *ctl->sqtail);

  n = *ctl->sqtail - *ctl->sqhead;
  do
    rc = uv__io_uring_enter(ctl->ringfd, n, n, UV__IORING_ENTER_GETEVENTS);
  while (rc == -1 && errno == EINTR);

  if (rc < 0)
    perror("libuv: io_uring_enter(getevents)");  /* Can't happen. */

  if (rc != (int) n)
    abort();

  assert(*ctl->sqhead == *ctl->sqtail);

  memcpy(oldevents, *events, sizeof(*events));

  /* Failed submissions are either EPOLL_CTL_DEL commands for file descriptors
   * that have been closed, or EPOLL_CTL_ADD commands for file descriptors
   * that we are already watching. Ignore the former and retry the latter
   * with EPOLL_CTL_MOD.
   */
  while (*ctl->cqhead != *ctl->cqtail) {
    slot = (*ctl->cqhead)++ & ctl->cqmask;

    cqe = ctl->cqe;
    cqe = &cqe[slot];

    if (cqe->res == 0)
      continue;

    fd = cqe->user_data >> 32;
    op = 3 & cqe->user_data;
    oldslot = 255 & (cqe->user_data >> 2);

    if (op == EPOLL_CTL_DEL)
      continue;

    if (op != EPOLL_CTL_ADD)
      abort();

    if (cqe->res != -EEXIST)
      abort();

    uv__epoll_ctl_prep(epollfd,
                       ctl,
                       events,
                       EPOLL_CTL_MOD,
                       fd,
                       &oldevents[oldslot]);
  }
}
```
</details>

I also tried to use the nats library without the libuv event loop, but then [I got a race condition when calling openssl functions inside the nats library](https://s3.amazonaws.com/clickhouse-test-reports/69772/e72843beb3cbcb06d2d461c7c421286ff618938b/integration_tests__tsan__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log)
<details>
  <summary>Example of sanitizer output</summary>

```
E           Exception: Sanitizer assert found for instance ==================
E           WARNING: ThreadSanitizer: data race (pid=1)
E             Write of size 4 at 0x728800018068 by thread T2 (mutexes: write M0, write M1, write M2):
E               #0 ssl3_write_bytes build_docker/./contrib/openssl/ssl/record/rec_layer_s3.c:282:16 (clickhouse+0x214018d9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #1 ssl3_write build_docker/./contrib/openssl/ssl/s3_lib.c:4510:12 (clickhouse+0x21396c1c) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #2 ssl_write_internal build_docker/./contrib/openssl/ssl/ssl_lib.c (clickhouse+0x213a7fb9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #3 SSL_write build_docker/./contrib/openssl/ssl/ssl_lib.c:2614:11 (clickhouse+0x213a81b6) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #4 natsSock_Write build_docker/./contrib/nats-io/src/comsock.c:494:21 (clickhouse+0x1e76f3d4) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #5 natsSock_WriteFully build_docker/./contrib/nats-io/src/comsock.c:582:13 (clickhouse+0x1e76f630) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #6 natsConn_bufferFlush build_docker/./contrib/nats-io/src/conn.c:292:13 (clickhouse+0x1e764881) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #7 natsConn_flushOrKickFlusher build_docker/./contrib/nats-io/src/conn.c:1746:13 (clickhouse+0x1e766760) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #8 natsConn_subscribeImpl build_docker/./contrib/nats-io/src/conn.c:3034:21 (clickhouse+0x1e7666c6) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #9 natsConnection_QueueSubscribe build_docker/./contrib/nats-io/src/sub.c:561:9 (clickhouse+0x1e789ae7) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #10 DB::NATSConsumer::subscribe() build_docker/./src/Storages/NATS/NATSConsumer.cpp:46:23 (clickhouse+0x19b3c789) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #11 DB::StorageNATS::subscribeConsumers() build_docker/./src/Storages/NATS/StorageNATS.cpp:235:23 (clickhouse+0x19aa82af) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #12 DB::StorageNATS::startup() build_docker/./src/Storages/NATS/StorageNATS.cpp:396:52 (clickhouse+0x19aaadf3) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #13 DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::__1::unique_ptr<DB::DDLGuard, std::__1::default_delete<DB::DDLGuard>>&, DB::LoadingStrictnessLevel) build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:1932:10 (clickhouse+0x17c67a22) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #14 DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:1671:20 (clickhouse+0x17c60b00) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #15 DB::InterpreterCreateQuery::execute() build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:2207:16 (clickhouse+0x17c6dbda) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #16 DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/./src/Interpreters/executeQuery.cpp:1327:40 (clickhouse+0x185448cd) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #17 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) build_docker/./src/Interpreters/executeQuery.cpp:1494:26 (clickhouse+0x1853fc37) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #18 DB::TCPHandler::runImpl() build_docker/./src/Server/TCPHandler.cpp:587:54 (clickhouse+0x1a531fb6) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #19 DB::TCPHandler::run() build_docker/./src/Server/TCPHandler.cpp:2520:9 (clickhouse+0x1a550507) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #20 Poco::Net::TCPServerConnection::start() build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x1eab3742) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #21 Poco::Net::TCPServerDispatcher::run() build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115:38 (clickhouse+0x1eab3fb1) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #22 Poco::PooledThread::run() build_docker/./base/poco/Foundation/src/ThreadPool.cpp:205:14 (clickhouse+0x1ea2cc22) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #23 Poco::(anonymous namespace)::RunnableHolder::run() build_docker/./base/poco/Foundation/src/Thread.cpp:45:11 (clickhouse+0x1ea2aeef) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #24 Poco::ThreadImpl::runnableEntry(void*) build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:335:27 (clickhouse+0x1ea293c9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E           
E             Previous write of size 4 at 0x728800018068 by thread T668:
E               #0 ssl3_read_bytes build_docker/./contrib/openssl/ssl/record/rec_layer_s3.c:674:16 (clickhouse+0x21402723) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #1 ssl3_read_internal build_docker/./contrib/openssl/ssl/s3_lib.c:4528:9 (clickhouse+0x21396ef9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #2 ssl3_read build_docker/./contrib/openssl/ssl/s3_lib.c:4551:12 (clickhouse+0x21396d72) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #3 ssl_read_internal build_docker/./contrib/openssl/ssl/ssl_lib.c (clickhouse+0x213a79d9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #4 SSL_read build_docker/./contrib/openssl/ssl/ssl_lib.c:2328:11 (clickhouse+0x213a7ed6) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #5 natsSock_Read build_docker/./contrib/nats-io/src/comsock.c:410:25 (clickhouse+0x1e76f155) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #6 _readLoop build_docker/./contrib/nats-io/src/conn.c:2212:13 (clickhouse+0x1e76d67d) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #7 _threadStart build_docker/./contrib/nats-io/src/unix/thread.c:39:5 (clickhouse+0x1e791512) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E           
E             Location is heap block of size 5408 at 0x728800018000 allocated by thread T2:
E               #0 malloc <null> (clickhouse+0x76526ac) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #1 CRYPTO_malloc build_docker/./contrib/openssl/crypto/mem.c:202:11 (clickhouse+0x215505ed) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #2 CRYPTO_zalloc build_docker/./contrib/openssl/crypto/mem.c:222:11 (clickhouse+0x215505ed)
E               #3 ossl_ssl_connection_new_int build_docker/./contrib/openssl/ssl/ssl_lib.c:728:9 (clickhouse+0x213a33d5) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #4 ossl_ssl_connection_new build_docker/./contrib/openssl/ssl/ssl_lib.c:927:12 (clickhouse+0x213a42a6) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #5 SSL_new build_docker/./contrib/openssl/ssl/ssl_lib.c:692:12 (clickhouse+0x213a30f9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #6 _makeTLSConn build_docker/./contrib/nats-io/src/conn.c:669:15 (clickhouse+0x1e76b8b4) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #7 _checkForSecure build_docker/./contrib/nats-io/src/conn.c:783:13 (clickhouse+0x1e76b8b4)
E               #8 _processExpectedInfo build_docker/./contrib/nats-io/src/conn.c:812:13 (clickhouse+0x1e76b8b4)
E               #9 _processConnInit build_docker/./contrib/nats-io/src/conn.c:1909:9 (clickhouse+0x1e76b8b4)
E               #10 _connect build_docker/./contrib/nats-io/src/conn.c:2000:21 (clickhouse+0x1e76928b) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #11 natsConnection_Connect build_docker/./contrib/nats-io/src/conn.c:3259:13 (clickhouse+0x1e768bdc) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #12 DB::NATSConnection::connectImpl() build_docker/./src/Storages/NATS/NATSConnection.cpp:150:25 (clickhouse+0x19ab8e14) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #13 DB::NATSConnection::connect() build_docker/./src/Storages/NATS/NATSConnection.cpp:105:5 (clickhouse+0x19ab8d2c) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #14 DB::StorageNATS::StorageNATS(DB::StorageID const&, std::__1::shared_ptr<DB::Context const>, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>, DB::LoadingStrictnessLevel) build_docker/./src/Storages/NATS/StorageNATS.cpp:109:39 (clickhouse+0x19aa4e5b) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #15 DB::StorageNATS* std::__1::construct_at[abi:v15007]<DB::StorageNATS, DB::StorageID const&, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>, DB::LoadingStrictnessLevel const&, DB::StorageNATS*>(DB::StorageNATS*, DB::StorageID const&, std::__1::shared_ptr<DB::Context>&&, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>&&, DB::LoadingStrictnessLevel const&) build_docker/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:35:48 (clickhouse+0x19ab6934) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #16 void std::__1::allocator_traits<std::__1::allocator<DB::StorageNATS>>::construct[abi:v15007]<DB::StorageNATS, DB::StorageID const&, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>, DB::LoadingStrictnessLevel const&, void, void>(std::__1::allocator<DB::StorageNATS>&, DB::StorageNATS*, DB::StorageID const&, std::__1::shared_ptr<DB::Context>&&, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>&&, DB::LoadingStrictnessLevel const&) build_docker/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:298:9 (clickhouse+0x19ab6934)
E               #17 std::__1::__shared_ptr_emplace<DB::StorageNATS, std::__1::allocator<DB::StorageNATS>>::__shared_ptr_emplace[abi:v15007]<DB::StorageID const&, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>, DB::LoadingStrictnessLevel const&>(std::__1::allocator<DB::StorageNATS>, DB::StorageID const&, std::__1::shared_ptr<DB::Context>&&, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>&&, DB::LoadingStrictnessLevel const&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:292:9 (clickhouse+0x19ab6934)
E               #18 std::__1::shared_ptr<DB::StorageNATS> std::__1::allocate_shared[abi:v15007]<DB::StorageNATS, std::__1::allocator<DB::StorageNATS>, DB::StorageID const&, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>, DB::LoadingStrictnessLevel const&, void>(std::__1::allocator<DB::StorageNATS> const&, DB::StorageID const&, std::__1::shared_ptr<DB::Context>&&, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>&&, DB::LoadingStrictnessLevel const&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:953:55 (clickhouse+0x19ab6934)
E               #19 std::__1::shared_ptr<DB::StorageNATS> std::__1::make_shared[abi:v15007]<DB::StorageNATS, DB::StorageID const&, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>, DB::LoadingStrictnessLevel const&, void>(DB::StorageID const&, std::__1::shared_ptr<DB::Context>&&, DB::ColumnsDescription const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::unique_ptr<DB::NATSSettings, std::__1::default_delete<DB::NATSSettings>>&&, DB::LoadingStrictnessLevel const&) build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962:12 (clickhouse+0x19aafc89) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #20 DB::registerStorageNATS(DB::StorageFactory&)::$_0::operator()(DB::StorageFactory::Arguments const&) const build_docker/./src/Storages/NATS/StorageNATS.cpp:717:16 (clickhouse+0x19aafc89)
E               #21 decltype(std::declval<DB::registerStorageNATS(DB::StorageFactory&)::$_0&>()(std::declval<DB::StorageFactory::Arguments const&>())) std::__1::__invoke[abi:v15007]<DB::registerStorageNATS(DB::StorageFactory&)::$_0&, DB::StorageFactory::Arguments const&>(DB::registerStorageNATS(DB::StorageFactory&)::$_0&, DB::StorageFactory::Arguments const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x19aafc89)
E               #22 std::__1::shared_ptr<DB::IStorage> std::__1::__invoke_void_return_wrapper<std::__1::shared_ptr<DB::IStorage>, false>::__call<DB::registerStorageNATS(DB::StorageFactory&)::$_0&, DB::StorageFactory::Arguments const&>(DB::registerStorageNATS(DB::StorageFactory&)::$_0&, DB::StorageFactory::Arguments const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:470:16 (clickhouse+0x19aafc89)
E               #23 std::__1::__function::__default_alloc_func<DB::registerStorageNATS(DB::StorageFactory&)::$_0, std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()[abi:v15007](DB::StorageFactory::Arguments const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x19aafc89)
E               #24 std::__1::shared_ptr<DB::IStorage> std::__1::__function::__policy_invoker<std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::__call_impl<std::__1::__function::__default_alloc_func<DB::registerStorageNATS(DB::StorageFactory&)::$_0, std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>>(std::__1::__function::__policy_storage const*, DB::StorageFactory::Arguments const&) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x19aafc89)
E               #25 std::__1::__function::__policy_func<std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()[abi:v15007](DB::StorageFactory::Arguments const&) const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0x194df025) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #26 std::__1::function<std::__1::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::operator()(DB::StorageFactory::Arguments const&) const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1197:12 (clickhouse+0x194df025)
E               #27 DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, DB::LoadingStrictnessLevel) const build_docker/./src/Storages/StorageFactory.cpp:229:16 (clickhouse+0x194df025)
E               #28 DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::__1::unique_ptr<DB::DDLGuard, std::__1::default_delete<DB::DDLGuard>>&, DB::LoadingStrictnessLevel) build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:1850:42 (clickhouse+0x17c66e1b) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #29 DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:1671:20 (clickhouse+0x17c60b00) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #30 DB::InterpreterCreateQuery::execute() build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:2207:16 (clickhouse+0x17c6dbda) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #31 DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) build_docker/./src/Interpreters/executeQuery.cpp:1327:40 (clickhouse+0x185448cd) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #32 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) build_docker/./src/Interpreters/executeQuery.cpp:1494:26 (clickhouse+0x1853fc37) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #33 DB::TCPHandler::runImpl() build_docker/./src/Server/TCPHandler.cpp:587:54 (clickhouse+0x1a531fb6) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #34 DB::TCPHandler::run() build_docker/./src/Server/TCPHandler.cpp:2520:9 (clickhouse+0x1a550507) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #35 Poco::Net::TCPServerConnection::start() build_docker/./base/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x1eab3742) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #36 Poco::Net::TCPServerDispatcher::run() build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:115:38 (clickhouse+0x1eab3fb1) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #37 Poco::PooledThread::run() build_docker/./base/poco/Foundation/src/ThreadPool.cpp:205:14 (clickhouse+0x1ea2cc22) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #38 Poco::(anonymous namespace)::RunnableHolder::run() build_docker/./base/poco/Foundation/src/Thread.cpp:45:11 (clickhouse+0x1ea2aeef) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
E               #39 Poco::ThreadImpl::runnableEntry(void*) build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:335:27 (clickhouse+0x1ea293c9) (BuildId: a478cfd27f9d94c642441be31a891dcade7ad3ed)
```
</details>

#### If you have any other ideas, I'm ready to try to implement them.

Reply:
So we need to stick the uv loop into a single thread, right? Then maybe it is better to create its own thread instead of occupying one from the background thread pool. Maybe combining that with conditional sleeps: if there is not nothing to execute, then let's sleep for X milliseconds. I think that should be fine. What do you think?

Maybe don't start to implement it yet, I will ask somebody who has better understanding of threading in ClickHouse to make sure that is in line with the rest of ClickHouse. If I am wrong, it might be unnecessary work for you.

Reply:
> Maybe don't start to implement it yet, I will ask somebody who has better understanding of threading in ClickHouse to make sure that is in line with the rest of ClickHouse.

Ok, in the meantime I will try to update the nats io library to the latest version and check it with thread sanitizer.

Reply:
So we got a green light for having a separate thread for each NATS table. The only extra info I got is to use `ThreadFromGlobalPool` instead of plain C++ threads.